File: clients/Go/model/entity_properties.go
package model

import "io"
import . "aicup2020/stream"

type EntityProperties struct {
    Size int32
    BuildScore int32
    DestroyScore int32
    CanMove bool
    PopulationProvide int32
    PopulationUse int32
    MaxHealth int32
    InitialCost int32
    SightRange int32
    ResourcePerHealth int32
    Build *BuildProperties
    Attack *AttackProperties
    Repair *RepairProperties
}
func NewEntityProperties(size int32, buildScore int32, destroyScore int32, canMove bool, populationProvide int32, populationUse int32, maxHealth int32, initialCost int32, sightRange int32, resourcePerHealth int32, build *BuildProperties, attack *AttackProperties, repair *RepairProperties) EntityProperties {
    return EntityProperties {
        Size: size,
        BuildScore: buildScore,
        DestroyScore: destroyScore,
        CanMove: canMove,
        PopulationProvide: populationProvide,
        PopulationUse: populationUse,
        MaxHealth: maxHealth,
        InitialCost: initialCost,
        SightRange: sightRange,
        ResourcePerHealth: resourcePerHealth,
        Build: build,
        Attack: attack,
        Repair: repair,
    }
}
func ReadEntityProperties(reader io.Reader) EntityProperties {
    result := EntityProperties {}
    result.Size = ReadInt32(reader)
    result.BuildScore = ReadInt32(reader)
    result.DestroyScore = ReadInt32(reader)
    result.CanMove = ReadBool(reader)
    result.PopulationProvide = ReadInt32(reader)
    result.PopulationUse = ReadInt32(reader)
    result.MaxHealth = ReadInt32(reader)
    result.InitialCost = ReadInt32(reader)
    result.SightRange = ReadInt32(reader)
    result.ResourcePerHealth = ReadInt32(reader)
    if ReadBool(reader) {
        var BuildValue BuildProperties
        BuildValue = ReadBuildProperties(reader)
        result.Build = &BuildValue
    } else {
        result.Build = nil
    }
    if ReadBool(reader) {
        var AttackValue AttackProperties
        AttackValue = ReadAttackProperties(reader)
        result.Attack = &AttackValue
    } else {
        result.Attack = nil
    }
    if ReadBool(reader) {
        var RepairValue RepairProperties
        RepairValue = ReadRepairProperties(reader)
        result.Repair = &RepairValue
    } else {
        result.Repair = nil
    }
    return result
}
func (value EntityProperties) Write(writer io.Writer) {
    WriteInt32(writer, value.Size)
    WriteInt32(writer, value.BuildScore)
    WriteInt32(writer, value.DestroyScore)
    WriteBool(writer, value.CanMove)
    WriteInt32(writer, value.PopulationProvide)
    WriteInt32(writer, value.PopulationUse)
    WriteInt32(writer, value.MaxHealth)
    WriteInt32(writer, value.InitialCost)
    WriteInt32(writer, value.SightRange)
    WriteInt32(writer, value.ResourcePerHealth)
    if value.Build == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        (*value.Build).Write(writer)
    }
    if value.Attack == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        (*value.Attack).Write(writer)
    }
    if value.Repair == nil {
        WriteBool(writer, false)
    } else {
        WriteBool(writer, true)
        (*value.Repair).Write(writer)
    }
}

File: clients/Python/model/entity_properties.py
from .build_properties import BuildProperties
from .attack_properties import AttackProperties
from .repair_properties import RepairProperties
class EntityProperties:
    def __init__(self, size, build_score, destroy_score, can_move, population_provide, population_use, max_health, initial_cost, sight_range, resource_per_health, build, attack, repair):
        self.size = size
        self.build_score = build_score
        self.destroy_score = destroy_score
        self.can_move = can_move
        self.population_provide = population_provide
        self.population_use = population_use
        self.max_health = max_health
        self.initial_cost = initial_cost
        self.sight_range = sight_range
        self.resource_per_health = resource_per_health
        self.build = build
        self.attack = attack
        self.repair = repair
    @staticmethod
    def read_from(stream):
        size = stream.read_int()
        build_score = stream.read_int()
        destroy_score = stream.read_int()
        can_move = stream.read_bool()
        population_provide = stream.read_int()
        population_use = stream.read_int()
        max_health = stream.read_int()
        initial_cost = stream.read_int()
        sight_range = stream.read_int()
        resource_per_health = stream.read_int()
        if stream.read_bool():
            build = BuildProperties.read_from(stream)
        else:
            build = None
        if stream.read_bool():
            attack = AttackProperties.read_from(stream)
        else:
            attack = None
        if stream.read_bool():
            repair = RepairProperties.read_from(stream)
        else:
            repair = None
        return EntityProperties(size, build_score, destroy_score, can_move, population_provide, population_use, max_health, initial_cost, sight_range, resource_per_health, build, attack, repair)
    def write_to(self, stream):
        stream.write_int(self.size)
        stream.write_int(self.build_score)
        stream.write_int(self.destroy_score)
        stream.write_bool(self.can_move)
        stream.write_int(self.population_provide)
        stream.write_int(self.population_use)
        stream.write_int(self.max_health)
        stream.write_int(self.initial_cost)
        stream.write_int(self.sight_range)
        stream.write_int(self.resource_per_health)
        if self.build is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.build.write_to(stream)
        if self.attack is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.attack.write_to(stream)
        if self.repair is None:
            stream.write_bool(False)
        else:
            stream.write_bool(True)
            self.repair.write_to(stream)
    def __repr__(self):
        return "EntityProperties(" + \
            repr(self.size) + "," + \
            repr(self.build_score) + "," + \
            repr(self.destroy_score) + "," + \
            repr(self.can_move) + "," + \
            repr(self.population_provide) + "," + \
            repr(self.population_use) + "," + \
            repr(self.max_health) + "," + \
            repr(self.initial_cost) + "," + \
            repr(self.sight_range) + "," + \
            repr(self.resource_per_health) + "," + \
            repr(self.build) + "," + \
            repr(self.attack) + "," + \
            repr(self.repair) + \
            ")"
